File: others/2000-2100/2092. 找出知晓秘密的所有专家.py
from math import inf
from typing import Dict, List
from heapq import *


class Solution:
    def findAllPeople(
        self, n: int, meetings: List[List[int]], firstPerson: int
    ) -> List[int]:
        graph = [[] for _ in range(n)]
        for a, b, t in meetings:
            graph[a].append((b, t))
            graph[b].append((a, t))

        h = []
        heapify(h)
        heappush(h, (0, 0))
        heappush(h, (0, firstPerson))
        vis = [inf] * n
        vis[0] = 0
        vis[firstPerson] = 0
        while h:
            [t, person] = heappop(h)

            for v, t2 in graph[person]:
                if t > t2:
                    continue
                if vis[v] == inf or t2 < vis[v]:
                    vis[v] = t2
                    heappush(h, (t2, v))
        ans = [i for i, x in enumerate(vis) if x != inf]
        return ans


class Solution:
    def findAllPeople(
        self, n: int, meetings: List[List[int]], firstPerson: int
    ) -> List[int]:
        # 将 meetings 按时间排序
        meetings.sort(key=lambda x: x[2])

        # 根据时间先后，为每个时刻构建一个邻接表,置于列表中。考虑到本就是稀疏图并再次拆分，非常分散，故每个邻接表都用 hash map。
        multiAdj = []
        for i in range(len(meetings)):
            if i == 0 or meetings[i][2] > meetings[i - 1][2]:
                multiAdj.append({})
            adj = multiAdj[-1]
            u, v, time = meetings[i]
            if u not in adj:
                adj[u] = []
            if v not in adj:
                adj[v] = []
            adj[u].append(v)
            adj[v].append(u)

        visited = [False] * n
        visited[0] = True
        visited[firstPerson] = True

        def dfs(u: int, adj: Dict[int, List[int]]):
            for v in adj.get(u, []):
                if not visited[v]:
                    visited[v] = True
                    dfs(v, adj)

        for adj in multiAdj:
            # 选取访问过的点 dfs 向外扩散，此处切忌遍历 0～n-1
            for p in adj.keys():
                if visited[p]:
                    dfs(p, adj)

        # 筛选访问过的点构成结果，即所有知晓秘密的专家
        return [i for i in range(n) if visited[i]]

File: leetcode/2000-2100/2092. 找出知晓秘密的所有专家.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define i128 __int128_t
#define db long double
#define pb emplace_back
#define pf emplace_front
#define pob pop_back
#define ep emplace
#define ins insert
#define all(x) (x).begin(), (x).end()
#define all2(x, i) (x).begin() + (i), (x).end()
using pii = pair<ll, ll>;
#define ull unsigned long long
#define vi vector<int>
#define vp vector<pii>
#define vl vector<long long>
#define vvi vector<vector<int>>
#define vvp vector<vector<pii>>
#define vvl vector<vector<long long>>
#define D(i, j, k) for (int(i) = (j); (i) >= (k); (i)--)
#define SZ(a) ((int) (a).size())
#define prq priority_queue
#define fi first
#define se second
constexpr int MOD = int(1e9 + 7);
constexpr long long INF = 0x3f3f3f3f3f3f3f3f;
constexpr int inf = 0x3f3f3f3f;
#define F(i, j, k) for (int(i) = (j); (i) <= (k); (i)++)

class Solution {
public:
    vector<int> findAllPeople(int n, vector<vector<int>> &meetings, int firstPerson) {
        sort(meetings.begin(), meetings.end(), [](const auto &a, const auto &b) { return a[2] < b[2]; });

        vector<bool> vis(n, false);
        vis[0] = true;
        vis[firstPerson] = true;

        int m = meetings.size();
        int i = 0;

        while (i < m) {
            int j = i;
            int time = meetings[i][2];

            while (j < m && meetings[j][2] == time) {
                j++;
            }

            unordered_map<int, vector<int>> adj;
            unordered_set<int> st;

            for (int k = i; k < j; k++) {
                int u = meetings[k][0];
                int v = meetings[k][1];
                adj[u].push_back(v);
                adj[v].push_back(u);
                st.insert(u);
                st.insert(v);
            }

            queue<int> q;
            for (int p: st) {
                if (vis[p]) {
                    q.push(p);
                }
            }

            while (!q.empty()) {
                int u = q.front();
                q.pop();

                for (int v: adj[u]) {
                    if (!vis[v]) {
                        vis[v] = true;
                        q.push(v);
                    }
                }
            }

            i = j;
        }

        vector<int> ans;
        for (int k = 0; k < n; k++) {
            if (vis[k])
                ans.push_back(k);
        }
        return ans;
    }
};
